### ISSUE TITLE
overload stringify 5.18-5.22

### ISSUE THREAD

**Issue by atoomic** [rurban/perl-compiler#219]
Env:
perl 5.20.2
B::C using bc522 with HEAD @734e776 ( cannot use 23016e1 with 5.20 )

Steps to reproduce

````
> cat test.pl
package OverloadTest;
use overload '""' => sub { ${$_[0]} };

package main;
my $foo = bless \(my $bar = "ok"), "OverloadTest";
print $foo."\n";

> perl test.pl
ok
> perlcc -r test.pl
pccOxMXJ.c:8483: warning: initialization makes pointer from integer without a cast
....
````

Seems a regression as this is working on 5.14

**Comment by rurban** [comment]
This started failing with 5.18.

I think the root-cause is my wokraround for the broken %version:: SvSTASH. See https://rt.cpan.org/Ticket/Display.html?id=81635

**Comment by toddr** [comment]
Right. You reported the bug against version except Devel::Peek and version are both perl as upstream. https://rt.cpan.org/Public/Bug/Display.html?id=107459

**Comment by toddr** [comment]
Bisected in blead.

c07f9fb2c73bebcf70fabbf464c0c3452af4fcbf is the first bad commit
commit c07f9fb2c73bebcf70fabbf464c0c3452af4fcbf
Author: Father Chrysostomos <sprout@cpan.org>
Date:   Tue Nov 20 13:47:27 2012 -0800

    Revert "Revert "8c34e50dc slowed down detruction with no DESTROY""
    
    This reverts commit 95f9781bc2fad025553db0160ef9c2c5363312a1.
    
    Now that the crash has been fixed by the preceding commit, we can
    reinstate 7cc6787e9db.

:100644 100644 2d1d887fe8b4256b2b5682dc6d26065f12ab5a6d be2038f4e7cdaf2d918c5df419348fca09290ab0 M	mro.c
:100644 100644 9f5c157a87b3f41f96566337e612a3674c129bbd a7914988e7861ead362d718960a408fd5bd0a5e3 M	sv.c
bisect run success

**Comment by toddr** [comment]
Output at offending commit:

cd t && (rm -f perl; /bin/ln -s ../perl perl)
SV = IV(0x8c1e234) at 0x8c1e238
  REFCNT = 1
  FLAGS = (TEMP,ROK)
  RV = 0x8c1e418
  SV = PVHV(0x8c24010) at 0x8c1e418
    REFCNT = 4
    FLAGS = (OOK,OVERLOAD,SHAREKEYS)
    STASH = 0x10/root/bisect.sh: line 10: 26914 Segmentation fault      ./perl -Ilib -MDevel::Peek=Dump -E'version->new("v5.22.0");Dump(\%version::)'

**Comment by toddr** [comment]
Reported to upstream. https://rt.perl.org/Public/Bug/Display.html?id=126410

**Comment by toddr** [comment]
Even with the patch applied to upstream, the original problem continues to be an issue so I think the upstream Devel::Peek regression is probably a distraction from the primary issue.

**Comment by toddr** [comment]
Test results with fix sent to upstream compiled in to perl 5.22.0

$>perlcc -e'package OverloadTest; use overload q{""} => sub { ${$_[0]} }; package main; my $foo = bless \(my $bar = "ok"), "OverloadTest"; print $foo."\n";' ; ./a.out
./a.out
/usr/local/cpanel/3rdparty/perl/522-debug/bin/perlcc: Unexpected compiler output
[WARNING] skip SvSTASH for overloaded HV %Errno:: flags=0x3280000c
 [WARNING] unknown (HV*)&sv_list[455] found in @static_free
OverloadTest=SCALAR(0x8112fb0)

**Comment by toddr** [comment]
Todd to bisect this today.

**Comment by rurban** [comment]
Patched my_curse for -fno-destruct (-O3) so we don't even need the core patch on older perls.
See bd78fe8

**Comment by toddr** [comment]
```git
$>git bisect bad
bee7c5743fabd49c584c5edbb637c66c5c25ca9a is the first bad commit
commit bee7c5743fabd49c584c5edbb637c66c5c25ca9a
Author: Father Chrysostomos <sprout@cpan.org>
Date:   Fri May 18 17:02:39 2012 -0700

    sv.c: Don’t fiddle with AMAGIC in sv_bless
    
    Since overloading itself now checks whether caches are up to date, and
    since changes to the stash (@ISA, methods) turn the flag on and over-
    loading itself turns the flag off when it can, sv_bless no longer
    needs to deal with it at all.

:040000 040000 4a9679f7a036de75ae61b9dda7ebecc3bf5335ba 634b1a9b9c56600fac5f9d9e4335474d106cd95e M	lib
:100644 100644 0c940cb434f461f262ab8263d2f9f1552a04bcf2 1e7f4d2e341d79f5306194808fd65530a1f628f6 M	sv.c
```

**Comment by rurban** [comment]
Uh, now that's interesting and should be pretty simple to fix, thanks!

**Comment by toddr** [comment]
./t/testc.sh 219

**Comment by toddr** [comment]
https://github.com/CpanelInc/perl-compiler/issues/27 needed a fix because overload cache initialiazation was needed for IO::File. This broke because $pkg was a B::CV and didn't implement NAME. 

It's possible that our %Exporter:: patch triggered the error and may be part of the solution?

**Comment by atoomic** [comment]
a workaround for it is to adjust ISA at run time (using  the overload branch)
```perl
#!perl 

package OverloadTest;
use overload '""' => sub { ${$_[0]} };

package main;
my $foo = bless \(my $bar = "ok" ), "OverloadTest";

INIT {
	push @OverloadTest::ISA, 'overload';
}
print "$foo\n";
````

**Comment by toddr** [comment]
This is because the HvMROMETA (xpvhv_aux) is updated for the stash when @ISA is changed. We need to start capturing and saving xpvhv_aux for all stashes for now on. This also would probably take away the need to set 'c3' on classes during XS init since this is simply populating that data.

**Comment by atoomic** [comment]
this dirty commit fixes the overload
https://github.com/CpanelInc/perl-compiler/commit/356e9e152e0b12cea2ee3fdd5baa56024844fc9f

**Comment by toddr** [comment]
@rurban While that does fix things, the correct solution is to save and properly attach xpvhv_aux where necessary. From the looks of it, the data is static?

**Comment by rurban** [comment]
No, it's dynamic. I only save it for RITER/EITER, see there.
Testing in branch `debug-stashcv-gh219` dbfa262840c2f60c084bc5cb887575258c67d7b0

**Comment by rurban** [comment]
Fixed with 1.52_18, thanks @atoomic
